Slight improvement of small_vector performance

hkaiser opened the pull request:
- adding benchmark

msimberg commented on libs/core/datastructures/include/hpx/datastructures/detail/small_vector.hpp:
```diff
@@ -48,9 +48,11 @@ namespace hpx::detail {
         void* do_allocate(std::size_t count, std::size_t) override
         {
-            return allocator_.allocate(count);
+            return allocator_.allocate((count + sizeof(T) - 1) / sizeof(T));
```
This is the key change, right?

hkaiser replied:
This is not the actual problem but reduces the memory allocation (if it happens) to a minimum. Before it was allocating too much memory on expansion.

msimberg replied:
Ah, right. This actually reduces the amount of memory allocated.

msimberg commented:
"Slight improvement": does that mean that Boost's implementation is still faster in the benchmark, or something else?

hkaiser commented:
> "Slight improvement": does that mean that Boost's implementation is still faster in the benchmark, or something else?

That means that for me (on WIndows) Boost is still faster. I have not tested gcc's or clangs' implementation of `std::memory_resource`. I have added a benchmark to this PR, would you be able to run it in your environment?

To explain what I changed: the PR essentially removes an explicit default construction for the `std::aligned_storage_t<>` that holds the pre-allocated memory. We've had a similar issue with our `cache_line_data` in the past, so I'm hoping that my fix will do the trick.

msimberg commented:
> That means that for me (on Windows) Boost is still faster. I have not tested gcc's or clang's implementation of `std::memory_resource`. I have added a benchmark to this PR, would you be able to run it in your environment?

For the benchmark itself it unfortunately doesn't seem to make a big difference. This is with `tcmalloc`. I haven't tried on DLA-Future yet.

Before:
```
----------------Parameters---------------------
Vector size                       :       10
Number of tests                   :   100000
Display time in                   : Seconds
-----Average-(boost::small_vector<i, 1>)---- 
Average execution time :  5.5e-08
-----Average-(hpx::small_vector<i, 1>)------ 
Average execution time :  7.5e-08
-----Average-(boost::small_vector<i, 2>)---- 
Average execution time :  5.6e-08
-----Average-(hpx::small_vector<i, 2>)------ 
Average execution time :  5.7e-08
-----Average-(boost::small_vector<i, 4>)---- 
Average execution time :  3.5e-08
-----Average-(hpx::small_vector<i, 4>)------ 
Average execution time :  4.9e-08
-----Average-(boost::small_vector<i, 8>)---- 
Average execution time :  1.7e-08
-----Average-(hpx::small_vector<i, 8>)------ 
Average execution time :  3.1e-08
-----Average-(boost::small_vector<i, 16>)---- 
Average execution time :    7e-09
-----Average-(hpx::small_vector<i, 16>)------ 
Average execution time :  1.8e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)---- 
Average execution time : 1.93e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)------ 
Average execution time : 1.83e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)---- 
Average execution time : 1.82e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)------ 
Average execution time : 1.72e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)---- 
Average execution time : 1.56e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)------ 
Average execution time : 1.45e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)---- 
Average execution time :  9.4e-08
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)------ 
Average execution time : 1.14e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)---- 
Average execution time :  6.2e-08
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)------ 
Average execution time :  7.8e-08

```

After:
```
----------------Parameters---------------------
Vector size                       :       10
Number of tests                   :   100000
Display time in                   : Seconds
-----Average-(boost::small_vector<i, 1>)---- 
Average execution time :  5.5e-08
-----Average-(hpx::small_vector<i, 1>)------ 
Average execution time :  7.4e-08
-----Average-(boost::small_vector<i, 2>)---- 
Average execution time :  5.5e-08
-----Average-(hpx::small_vector<i, 2>)------ 
Average execution time :  5.6e-08
-----Average-(boost::small_vector<i, 4>)---- 
Average execution time :  3.5e-08
-----Average-(hpx::small_vector<i, 4>)------ 
Average execution time :  4.7e-08
-----Average-(boost::small_vector<i, 8>)---- 
Average execution time :  1.6e-08
-----Average-(hpx::small_vector<i, 8>)------ 
Average execution time :  2.9e-08
-----Average-(boost::small_vector<i, 16>)---- 
Average execution time :    7e-09
-----Average-(hpx::small_vector<i, 16>)------ 
Average execution time :  1.6e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)---- 
Average execution time : 1.99e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)------ 
Average execution time : 1.89e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)---- 
Average execution time : 1.85e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)------ 
Average execution time :  1.7e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)---- 
Average execution time : 1.52e-07
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)------ 
Average execution time : 1.47e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)---- 
Average execution time :  9.5e-08
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)------ 
Average execution time : 1.13e-07
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)---- 
Average execution time :  6.4e-08
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)------ 
Average execution time :  6.9e-08

```

hkaiser commented:
Thanks @msimberg. Back to the drawing board, then.

hkaiser commented:
@msimberg could you rerun the benchmark with `--vector_size=1`, please (also perhaps increase `--test_count` to 1e6/1e7)? That will push_back one element into each instance of the tested vectors. This is very close to what happens in hpx::future, I believe. 

The benchmarks you ran add 10 elements, which guarantees allocation (except for the `N==16` vectors). In those cases, both versions are very similar as the time is dominated by the allocations anyways.

From what I see from your timings above, the case without allocations has improved and has come closer to the boost vector.

hkaiser commented:
@msimberg FWIW, at least two of the performance tests report significant improvements.

msimberg commented:
> @msimberg could you rerun the benchmark with `--vector_size=1`, please (also perhaps increase `--test_count` to 1e6/1e7)? That will push_back one element into each instance of the tested vectors. This is very close to what happens in hpx::future, I believe.

Good point! There's still a gap to Boost, but this is definitely an improvement. Thanks for looking into this!

Before:
```
> ./bin/small_vector_benchmark_test --vector_size=1 --test_count=10000000
----------------Parameters---------------------
Vector size                       :        1
Number of tests                   : 10000000
Display time in                   : Seconds
-----Average-(boost::small_vector<i, 1>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 1>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 2>)---- 
Average execution time :        0
-----Average-(hpx::small_vector<i, 2>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 4>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 4>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 8>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 8>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 16>)---- 
Average execution time :        0
-----Average-(hpx::small_vector<i, 16>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)------ 
Average execution time :  1.6e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)---- 
Average execution time :    7e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)------ 
Average execution time :  1.5e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)---- 
Average execution time :    7e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)------ 
Average execution time :  1.6e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)------ 
Average execution time :  1.8e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)------ 
Average execution time :  2.4e-08
```
After:
```
> ./bin/small_vector_benchmark_test --vector_size=1 --test_count=10000000
----------------Parameters---------------------
Vector size                       :        1
Number of tests                   : 10000000
Display time in                   : Seconds
-----Average-(boost::small_vector<i, 1>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 1>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 2>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 2>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 4>)---- 
Average execution time :        0
-----Average-(hpx::small_vector<i, 4>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 8>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 8>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<i, 16>)---- 
Average execution time :    1e-09
-----Average-(hpx::small_vector<i, 16>)------ 
Average execution time :    9e-09
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 1>)------ 
Average execution time :  1.4e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 2>)------ 
Average execution time :  1.4e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 4>)------ 
Average execution time :  1.4e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 8>)------ 
Average execution time :  1.4e-08
-----Average-(boost::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)---- 
Average execution time :    6e-09
-----Average-(hpx::small_vector<N3hpx4util15unique_functionIFvvELb0EEE, 16>)------ 
Average execution time :  1.3e-08
```
